notebooks/02_model/hypertune_spark_deep_dive.ipynb

```diff
@@ -0,0 +1,983 @@
+      ]
+     },
+     "execution_count": 11,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "model.explainParams()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "[1.3818693574087881,\n",
+       " 1.03881169247089,\n",
+       " 1.4149480500849774,\n",
+       " 1.8957753237837804,\n",
+       " 2.707401263212467,\n",
+       " 1.3818693597518183,\n",
+       " 1.022026009040587,\n",
+       " 1.4517896758192947,\n",
+       " 1.9079444576340328,\n",
+       " 2.786421555547063]"
+      ]
+     },
+     "execution_count": 12,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "model.validationMetrics"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To get the best model, just do"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "model_best_spark = model.bestModel"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 3 Hyperparameter tuning with `hyperopt` in sequential mode"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "`hyperopt` is an open source Python package that is designed for tuning parameters for generic function with any pre-defined loss. More information about `hyperopt` can be found [here](https://github.com/hyperopt/hyperopt).\n",
+    "\n",
+    "In `hyperopt`, an *objective* function is defined for optimizing the hyper parameters. In this case, the objective is similar to that in the Spark native construct situation, which is *to the RMSE metric for an ALS recommender*. Parameters of `rank` and `regParam` are used as hyperparameters. \n",
+    "\n",
+    "The objective function shown below demonstrates a RMSE loss for an ALS recommender. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Customize an objective function\n",
+    "def objective(params):\n",
```

Comment:
I would try to follow here single responsibility. I would suggest:
`def objective(params, model, dataset)` or alternatively `def objective(params, model, train, valid)`. At a high level, this function takes a model and a dataset and find the best parameters.

Reply:
It is a little bit against SRP. I would do the same but there is some restriction in parsing arguments to `objective` when calling `fmin` in `hyperopt` for tuning.